### PR TITLE
Use C++14 when compiling CHIP for esp32.

### PR DESCRIPTION
**Pull request by bzbarsky-apple**
I couldn't figure out where the -std=-gnu++11 in CXXFLAGS was coming
from in the esp32 build; it might be somewhere in the SDK.  So we just
filter it out, and manually add -std=gnu++14 instead.

 #### Problem
We're compiling some CHIP bits with `-std=gnu++11` in esp32 builds, but CHIP wants to use C++14 features.

 #### Summary of Changes
Use `-std=gnu++14` when doing the CHIP compile.

### CODE REVIEW

**Review by andy31415 — APPROVED** [review]

**Review by rwalker-apple — APPROVED** [review]

**Comment by bzbarsky-apple** [comment]
I'm not quite sure why there's a size increase from using C++14.  There are some random size increases in `mdns::Minimal::*` destructors, extra CSWITCH symbols, etc....  Maybe C++14 pins down some semantics more than C++11 does that force the extra code.

**Review by dhrishi — APPROVED** [review]

File: examples/all-clusters-app/esp32/Makefile [hunk 1]
```diff
@@ -26,7 +26,7 @@ EXTRA_COMPONENT_DIRS += $(PROJECT_PATH)/third_party/connectedhomeip/config/esp32
                         $(PROJECT_PATH)/../../common/QRCode \
                         $(PROJECT_PATH)/../../common/screen-framework \
 
-CXXFLAGS += -std=c++11 -Os -DLWIP_IPV6_SCOPES=0
+CXXFLAGS += -std=c++14 -Os -DLWIP_IPV6_SCOPES=0
```
Comment (dhrishi) [inline examples/all-clusters-app/esp32/Makefile:29]: @bzbarsky-apple  Do you want to make similar change in other ESP32 examples? `examples/temperature-measurement-app/esp32/`

**Comment by bzbarsky-apple** [comment]
> it comes from here: https://github.com/espressif/esp-idf/blob/release/v4.1/make/project.mk#L495

Thank you for the pointer!

File: examples/all-clusters-app/esp32/Makefile (hunk 1 above)
Comment (bzbarsky-apple) [reply examples/all-clusters-app/esp32/Makefile:29]: Yep, good idea.  Done.

**Review by saurabhst — APPROVED** [review]

**Pull request merged**